server/api/forecast_rotation.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package api
+
+import (
+	"github.com/mattermost/mattermost-plugin-solar-lottery/server/utils/bot"
+)
+
+type Forecast struct {
+	StartingShift                      int
+	NumShifts                          int
+	SampleSize                         int
+	CountErrFailedInsufficientForNeeds int
+	CountErrFailedInsufficientForSize  int
+	CountErrFailedSizeExceeded         int
+	NeedErrCounts                      map[string]int
+	ShiftErrCounts                     []int
+
+	// for each user (MattermostUserID) contains NumShifts probabilities, of
+	// the user being selected into the respective shift number. This is based
+	// on successful guesses only.
+	UserShiftCounts map[string][]int
+	UserCounts      map[string]int
+}
+
+func (api *api) ForecastRotation(rotation *Rotation, startingShiftNumber, numShifts, sampleSize int) (*Forecast, error) {
+	err := api.Filter(
+		withActingUserExpanded,
+		withRotationExpanded(rotation),
+	)
+	if err != nil {
+		return nil, err
+	}
+	logger := api.Logger.Timed().With(bot.LogContext{
+		"Location":            "api.Forecast",
+		"ActingUsername":      api.actingUser.MattermostUsername(),
+		"NumShifts":           numShifts,
+		"StartingShiftNumber": startingShiftNumber,
+		"RotationID":          rotation.RotationID,
+	})
+
+	f := &Forecast{
+		StartingShift:   startingShiftNumber,
+		NumShifts:       numShifts,
+		SampleSize:      sampleSize,
+		NeedErrCounts:   map[string]int{},
+		ShiftErrCounts:  make([]int, numShifts),
+		UserShiftCounts: map[string][]int{},
+		UserCounts:      map[string]int{},
+	}
+
+GUESS:
+	for i := 0; i < sampleSize; i++ {
+		var shifts []*Shift
+		// Guess' logs are too verbose - suppress
+		prevLogger := api.Logger
+		api.Logger = &bot.NilLogger{}
+		shifts, err = api.Guess(rotation, startingShiftNumber, numShifts)
+		api.Logger = prevLogger
+		var aerr autofillError
+		if err != nil {
+			var ok bool
+			aerr, ok = err.(autofillError)
+			if !ok {
+				return nil, err
+			}
+
+			switch aerr.orig {
+			case ErrFailedInsufficientForNeeds:
+				f.CountErrFailedInsufficientForNeeds++
```

Comment:
TODO: This is currently broken - `aerr` has 2 fields, one for the actual failing need, and another for all unmet needs; should be accounting for both, separately.